### DISCUSSION TITLE
toDataURL on Safari (iOS / Desktop) could lead to "exceeded memory error"

### DISCUSSION THREAD

**Discussion by and2** [fabricjs/fabric.js#7923] [Q&A]
Safari (desktop and iOS) is known to be a [canvas hoarder](https://pqina.nl/blog/total-canvas-memory-use-exceeds-the-maximum-limit/). Setting the **height** and the **width** of the canvas to **0** before removing the canvas element will get Safari to release the memory a lot quicker. 

We've experienced this issue especially on iOS 15 where the max memory for a new iPhone is 384MB. 
But could be replicated on desktop Safari as well.

The following has significantly improved the memory situation: 

```
fabric.util.object.extend(
    fabric.StaticCanvas.prototype,
    {
      toDataURL: function (options) {
        options || (options = { })
        const format = options.format || 'png'
        const quality = options.quality || 1
        const multiplier = (options.multiplier || 1) * (options.enableRetinaScaling ? this.getRetinaScaling() : 1)
        const canvasEl = this.toCanvasElement(multiplier, options)
        const data = fabric.util.toDataURL(canvasEl, format, quality)
        canvasEl.width = 0
        canvasEl.height = 0
        return data
      }
    }
  )
```

Noting here 
```
canvasEl.width = 0 
canvasEl.height = 0
```

before returning the data

Would be good to know 
1. If anyone suggests a better approach to this
2. If there is a way to find all the canvas elements created with `fabric.util.createCanvasElement()` so that before destroying the fabric canvas all these will also get a width and height of 0.

**Comment by ShaMan123** [comment]
This is an excellent PR.
I can back you.
